<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <script>
        function increase(number) {
            /* 비동기 결과(성공, 실패를 가진 객체(콜백 지옥을 피하기 위해 사용됨.) */
            const promise = new Promise((resolve, reject) => {
                setTimeout(() => {
                    const result = number + 10;
                    
                    if(result > 50) { // 50보다 크면 실패
                        const e  = new Error('NumberTooBig');

                        return reject(e);
                    }

                    return resolve(result); // 성공했을 때 resolve() 호출
                }, 1000); // 1초 후에 실행
            })

            return promise;
        }

        console.log(increase(0)); // Promise { <pending> }
        
            increase(0)
            .then(result => {
                console.log('성공결과:',result);
                return increase(result);
            })
            .then(result => {
                console.log('성공결과:',result);
                return increase(result);
            })
            .then(result => {
                console.log('성공결과:',result);
                return increase(result);
            })
            .then(result => {
                console.log('성공결과:',result);
                return increase(result);
            })
            .then(result => {
                console.log('성공결과:',result);
                return increase(result);
            })
            .then(result => {
                console.log('성공결과:',result);
                console.log('end');
                return increase(result);
            })
            .catch(error => console.log(error)) // 에러가 발생했을 때 catch() 호출
            .finally(() => console.log('반드시 실행되고 반드시 해야할 작업'));

            /* 
                then: 이전 작업이 성공적으로 완료 되었을때 promise에 담긴 성공 값(resolve)을 
                추출하여 확인하는 콜백 함수
                catch: 작업 중 에러가 발생했을 때 동작하는 메소드
                finally: 성공, 실패와 관계없이 반드시 실행되는 콜백 함수
                (finally는 then, catch 다음에 위치해야 한다.)
                (finally는 Promise 객체를 반환하지 않기 때문에 then, catch를 체이닝할 수 없다.)

                (이 메소드들을 통해 들여쓰기가 줄어들어 가독성이 좋아지고 유지보수성이 향상된다.)
            */
    </script>
</body>
</html>